odoo/upgrade_code/18.1-01-rename-class.py

```diff
@@ -0,0 +1,181 @@
+    ])
+
+
+def attribute_to_model_name(line):
+    if '(' in line or ',' in line:
+        return
+    if '[' in line:
+        # makes the code idempotent
+        return
+    if '"' not in line and "'" not in line:
+        return
+    if line.startswith('     ') or line.strip().startswith('#'):
+        return
+    if not line.split('=', 1)[0].strip() in ('_name', '_inherit'):
+        return
+    return re.sub(r'''[\]\['"\s\n]+''', '', line.split('=').pop())
+
+
+def get_class_name_models(file):
+    content = file.content
+    class_name_models = {}
+    iter_lines = iter(content.split('\n'))
+    line = next(iter_lines, None)
+    while line is not None:
+        part = regex_class_name.search(line)
+        line = next(iter_lines, None)
+
+        if not part:
+            continue
+
+        class_name = part.group(1)
+        current_model_name = None
+        while line is not None and (not line or line.startswith('  ')):
+            class_model_name = attribute_to_model_name(line)
+            if class_model_name and ('_name' in line or ('_inherit' in line and not current_model_name)):
+                current_model_name = class_model_name
+            line = next(iter_lines, None)
+
+        if current_model_name and current_model_name != 'base':
+            new_class_name = model_name_to_class_name(current_model_name)
+
+            previous = class_name_models.get(class_name)
+            if previous and previous[1] != current_model_name:
+                raise ValueError(f'Please fix your class {class_name!r} from addon {file.addon.name!r} used for {previous[1]!r} and {current_model_name!r} ({file.path})')
+
+            class_name_models[class_name] = (new_class_name, current_model_name)
+    return class_name_models
+
+
+def update_class_name(content, class_name_models):
+    # update class name
+    lines = content.split('\n')
+    for old_class_name, (new_class_name, _current_model_name) in class_name_models.items():
+        if old_class_name == new_class_name:
+            continue
+        if old_class_name == new_class_name:
+            # to avoid unwanted changes (classes having the table name)
+            continue
+
+        reg_use_class_name = re.compile(rf'\b{old_class_name}\b')
+        reg_use_super = re.compile(rf'super\(\s*{new_class_name}\s*,\s*self\s*\)')
+        for index, line in enumerate(lines):
+            try:
+                start = line.split(old_class_name, 1)[0]
+                if "#" in start or ' fields.' in start or start.endswith('self.') or start.count('"') % 2 or start.count("'") % 2:
+                    continue
+                if (" FROM " in line or
+                    "JOIN " in line or
+                    " UNION " in line or
+                    " AND " in line or
+                    " ON " in line or
+                    " OR " in line or
+                    " WHERE " in line or
+                    " GROUP BY " in line or
+                    " SELECT " in line):
```

Comment:
Not sure if a regex could be faster?

Reply:
Speed ​​is of no interest for a script executed once, on files, outside production.